triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala

```diff
@@ -384,50 +354,96 @@ class Runner(
+        case _ => {
+          val errMsg = s"Identifier ${triggerId.qualifiedName} does not point to a trigger"
+          throw new RuntimeException(errMsg)
+        }
+      }
+    val triggerIds = TriggerIds(tyCon.packageId)
+    if (tyCon == triggerIds.damlTriggerLowLevel("Trigger")) {
+      logger.debug("Running low-level trigger")
+      val triggerVal = EVal(triggerId)
+      val triggerTy = TypeConApp(tyCon, ImmArray(stateTy))
+      Trigger(triggerVal, triggerTy, triggerIds)
+    } else if (tyCon == triggerIds.damlTrigger("Trigger")) {
+      logger.debug("Running high-level trigger")
+      val runTrigger = EVal(triggerIds.damlTrigger("runTrigger"))
+      val triggerState = TTyCon(triggerIds.damlTriggerInternal("TriggerState"))
+      val lowLevelTriggerTy = triggerIds.damlTriggerLowLevel("Trigger")
+      val lowTriggerVal = EApp(runTrigger, EVal(triggerId))
+      val lowStateTy = TApp(triggerState, stateTy)
+      val lowTriggerTy = TypeConApp(lowLevelTriggerTy, ImmArray(lowStateTy))
+      Trigger(lowTriggerVal, lowTriggerTy, triggerIds)
+    } else {
+      val errMsg =
+        s"Identifier ${triggerId.qualifiedName} does not point to a trigger. Its type must be Daml.Trigger.Trigger or Daml.Trigger.LowLevel.Trigger."
+      throw new RuntimeException(errMsg)
+    }
+  }
+
+  // Construct a trigger runner from the given DAR.
+  def fromDar(
       dar: Dar[(PackageId, Package)],
       triggerId: Identifier,
       client: LedgerClient,
       timeProviderType: TimeProviderType,
       applicationId: ApplicationId,
-      party: String
-  )(implicit materializer: Materializer, executionContext: ExecutionContext): Future[SExpr] = {
-    val runner = new Runner(
+      party: String): Runner = {
+    val darMap = dar.all.toMap
+    val compiler = Compiler(darMap)
+    val compiledPackages =
+      PureCompiledPackages(darMap, compiler.compilePackages(darMap.keys)).right.get
+    val trigger = Runner.getTrigger(compiledPackages, triggerId)
+    new Runner(
       client,
+      timeProviderType,
       applicationId,
-      party,
-      dar
+      compiledPackages,
+      trigger,
+      party
     )
-    val (triggerExpr, triggerTy, triggerIds) = runner.getTrigger(triggerId)
-    val converter = Converter(runner.compiledPackages, triggerIds)
-    val filter = runner.getTriggerFilter(converter, triggerExpr, triggerTy)
-    val heartbeat = runner.getTriggerHeartbeat(converter, triggerExpr, triggerTy)
+  }
+
+  // Convience wrapper that creates the runner and runs the trigger.
+  def run(
+      dar: Dar[(PackageId, Package)],
+      triggerId: Identifier,
+      client: LedgerClient,
+      timeProviderType: TimeProviderType,
+      applicationId: ApplicationId,
+      party: String
+  )(implicit materializer: Materializer, executionContext: ExecutionContext): Future[SExpr] = {
+    val runner = Runner.fromDar(dar, triggerId, client, timeProviderType, applicationId, party)
+    val filter = runner.getTriggerFilter()
```

Comment:
I think the API isn’t quite right get. It would make more sense to split the initialization into
1. Take the packages and a trigger id, call `getTrigger`, `getTriggerFilter` and `getTriggerHeartbeat` and stuff that information in `Trigger`.
2. Then make the runner take the `Trigger` and only handle connection to the ledger.

However, I didn’t want to include this in this PR to make it easier to follow.